onnxruntime/core/providers/cuda/tensor/gather_elements.cc

```diff
@@ -1,101 +1,216 @@
+      default:
+        ORT_THROW("Unsupported indices element size by the GatherElements CUDA kernel");
+    }
+    return Status::OK();
+  }
+};
+
+#undef CASE_GATHER_ELEMENTS_IMPL
 
 Status GatherElements::ComputeInternal(OpKernelContext* context) const {
   // Process input data tensor
   const auto* input_tensor = context->Input<Tensor>(0);
   const auto& input_shape = input_tensor->Shape();
-  const auto& input_dims = input_shape.GetDims();
-  const int64_t input_rank = static_cast<int64_t>(input_dims.size());
+  const int64_t input_rank = static_cast<int64_t>(input_shape.NumDimensions());
 
   // Process indices tensor
   const auto* indices_tensor = context->Input<Tensor>(1);
   const auto& indices_shape = indices_tensor->Shape();
-  const auto& indices_dims = indices_shape.GetDims();
-  const int32_t indices_rank = static_cast<int32_t>(indices_dims.size());
   const int64_t indices_size = indices_shape.Size();
 
   // Handle negative axis if any
   const int64_t axis = HandleNegativeAxis(axis_, input_rank);
 
-  // Validate input shapes and ranks (invoke the static method in the CPU GatherElements kernel that hosts the shared checks)
-  auto status = onnxruntime::GatherElements::ValidateInputShapes(input_shape, indices_shape, axis);
-  if (!status.IsOK())
-    return status;
+  // Validate input shapes and ranks (invoke the static method in the CPU GatherElements kernel that hosts the shared
+  // checks)
+  ORT_RETURN_IF_ERROR(onnxruntime::GatherElements::ValidateInputShapes(input_shape, indices_shape, axis));
 
   // create output tensor
   auto* output_tensor = context->Output(0, indices_shape);
 
   // if there are no elements in 'indices' - nothing to process
-  if (indices_shape.Size() == 0)
-    return Status::OK();
+  if (indices_size == 0) return Status::OK();
 
-  // Set stride along axis to 0 so we don't need IF statment to check in kernel.
-  TensorPitches masked_input_strides(input_dims);
-  int64_t input_stride_along_axis = masked_input_strides[axis];
-  masked_input_strides[axis] = 0;
-  TArray<int64_t> gpu_masked_input_strides(masked_input_strides);
+  TensorShapeVector input_shape_vec = input_shape.AsShapeVector();
+  TensorShapeVector indices_shape_vec = indices_shape.AsShapeVector();
+  int64_t new_axis, new_rank, input_stride_along_axis;
+  TArray<int64_t> masked_input_strides;
+  TArray<fast_divmod> indices_fdms;
+  CoalesceDimensions(input_shape_vec, indices_shape_vec, axis, new_axis, new_rank, input_stride_along_axis,
+                     masked_input_strides, indices_fdms);
 
-  TArray<fast_divmod> fdm_indices_strides(indices_rank);
-  TensorPitches indices_strides(indices_dims);
-  for (auto i = 0; i < indices_rank; i++) {
-    fdm_indices_strides[i] = fast_divmod(gsl::narrow_cast<int>(indices_strides[i]));
+  // Use element size instead of concrete types so we can specialize less template functions to reduce binary size.
```

Comment:
you mean, compared with previous code, there will be smaller binary size, or we are trying to avoid bigger binary size when we start using MLTypeCallDispatcher in this PR?

Reply:
If you review the same part in scatter_elements.cc, this comment may more meaningful. Previously in scatter_elements.cc, we listed all possible types as below, but this is not necessary. Previous code in gather_elements.cc doesn't have this problem, but since after I changed the code, the code is similar, so I just put same comment in both places.

`utils::MLTypeCallDispatcher<float, MLFloat16, int16_t, int8_t, int32_t, int64_t, uint8_t, uint16_t, uint32_t, uint64_t, double, bool>`